src/main/java/org/folio/consortia/service/impl/UserAffiliationAsyncServiceImpl.java

```diff
@@ -0,0 +1,90 @@
+package org.folio.consortia.service.impl;
+
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.IntStream;
+
+import org.apache.commons.lang3.StringUtils;
+import org.folio.consortia.config.kafka.KafkaService;
+import org.folio.consortia.domain.dto.PrimaryAffiliationEvent;
+import org.folio.consortia.domain.dto.Tenant;
+import org.folio.consortia.domain.dto.User;
+import org.folio.consortia.domain.entity.TenantEntity;
+import org.folio.consortia.repository.UserTenantRepository;
+import org.folio.consortia.service.UserAffiliationAsyncService;
+import org.folio.consortia.service.UserService;
+import org.folio.consortia.service.UserTenantService;
+import org.springframework.stereotype.Service;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+
+import lombok.AllArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.extern.log4j.Log4j2;
+
+@Service
+@Log4j2
+@AllArgsConstructor
+public class UserAffiliationAsyncServiceImpl implements UserAffiliationAsyncService {
+
+  private final UserTenantService userTenantService;
+  private final KafkaService kafkaService;
+  private final UserService userService;
+  private final UserTenantRepository userTenantRepository;
+  private static final ObjectMapper OBJECT_MAPPER;
+
+  static {
+    OBJECT_MAPPER = new ObjectMapper()
+      .findAndRegisterModules()
+      .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+      .configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false)
+      .setSerializationInclusion(JsonInclude.Include.NON_EMPTY);
```

Comment:
Nice to have: I propose to add config bean with @ Primary annotation to be safe and inject it in all places or declare object mapper in application.yaml properties or declare utility class(together with  createPrimaryAffiliationEvent )to avoid duplication